_drafts/measuring-stack-usage.md

```diff
@@ -0,0 +1,361 @@
+
+This is a technique I applied recently to measure stack usage without modifying
+the target device code.
+
+The idea is to use GDB to single-step through the program, and to dump the stack
+pointer register, `$sp`, at every step. I used a small GDB Python script to do
+this work, a small snippet is shown below:
+
+```python
+# save the starting stack pointer value
+sp_start = int(gdb.parse_and_eval("$sp"))
+print("Starting SP: 0x{:08x}".format(sp_start))
+
+# record the lowest (memory index) value- this is the max stack usage for the
+# function. initialize it to the maximum addressable address.
+sp_min = 0xFFFFFFFF
+while True:
+    # single instruction step, so we capture _any_ sp change
+    gdb.execute("si")
+    sp = int(gdb.parse_and_eval("$sp"))
+    if sp < sp_min:
+        sp_min = sp
+    # the full implementation exits this loop when hitting a breakpoint
+    if hit_breakpoint():
+      break
+print("Min SP: 0x{:08x}".format(sp_min))
+print("Delta SP: {}".format(sp_start - sp_min))
+```
+
+The script is _very slow_ to run since it interrupts program flow at every
+instruction, but it gives a very accurate result.
+
+It can be enhanced to produce more data, for example recording the max usage for
+each frame in a call stack:
+
+```python
+# write the verbose data to a file
+with open("stack_usage.log", "w") as f:
+    sp_min = 0xFFFFFFFF
+    current_function = None
+    current_function_max_su = 0
+    while True:
+        gdb.execute("si")
+        sp = int(gdb.parse_and_eval("$sp"))
+        su = sp_start - sp
+        name = gdb.selected_frame().name()
+        # if we just changed function name, record the high water mark
+        # from the last function
+        if current_function is not None and name != current_function:
+            f.write(f"{current_function}|{current_function_max_su}\n")
+            current_function_max_su = 0
+        current_function = name
+        if su > current_function_max_su:
+            current_function_max_su = su
+        # f.write(f"{current_function}|{current_function_max_su}\n")
+        if sp < sp_min:
+            sp_min = sp
+        # check if we've reached a breakpoint
+        if hit_breakpoint:
+            print("Hit breakpoint at 0x{:08x}".format(pc))
+            break
+    f.write(f"{current_function}|{current_function_max_su}\n")
+```
+
+Plotting that usage can be interesting:
+
+```bash
+# using the 'plotext' python library to plot the raw trace data on the terminal
+# https://github.com/piccolomo/plotext
+❯ cut -d \| -f 2 sp-all.trace | \
+     python -c 'import plotext as plt; import sys; \
+       data = [int(x.strip()) for x in sys.stdin.readlines()]; \
+       plt.plot(data, marker="hd"); \
+       plt.title("Stack Usage"); \
+       plt.plot_size(80,25); \
+       plt.theme("pro"); \
+       plt.show()'
+
+                                     Stack Usage
+     ┌─────────────────────────────────────────────────────────────────────────┐
```

Comment:
this renders so badly on mobile SIGH. replace with png :(.
![image](https://github.com/memfault/interrupt/assets/2538614/1c4fafee-dc18-4ab1-880e-253ee28113cf)

Reply:
sigh switched to gnuplot too. RIP my one-liner.